### ISSUE TITLE
Code doesn't compile with error prone

### ISSUE THREAD

**Issue by davido** [google/error-prone#503]
Trying to run Error Prone 2.0.15 on Buck code base with Buck (with this PR applied: https://github.com/facebook/buck/pull/1047), it's failing to compile tests (all is fine without Error Prone):

```
  @SafeVarargs
  public static <T> Matcher<Collection<? extends T>> hasConsecutiveItems(
      Matcher<? super T>... matchers) {
    return hasConsecutiveItems(Arrays.asList(matchers)); // This is the line 86
  }
```

The error message is:

```
$ buck build --config java.error_prone_javac=true test/com/facebook/buck/testutil:testutil
::: '.nobuckcheck' file is present. Not updating buck.
/home/davido/projects/davido_buck/test/com/facebook/buck/testutil/HasConsecutiveItemsMatcher.java:86: error: incompatible types: inference variable T has incompatible bounds
    return hasConsecutiveItems(Arrays.asList(matchers));
                              ^
    upper bounds: T,java.lang.Object
    lower bounds: T,T,org.hamcrest.Matcher<? super T>
Errors: 1. Warnings: 0.

BUILD FAILED: //test/com/facebook/buck/testutil:testutil failed with exit code 1:
javac_jar
stderr: /home/davido/projects/davido_buck/test/com/facebook/buck/testutil/HasConsecutiveItemsMatcher.java:86: error: incompatible types: inference variable T has incompatible bounds
    return hasConsecutiveItems(Arrays.asList(matchers));
                              ^
    upper bounds: T,java.lang.Object
    lower bounds: T,T,org.hamcrest.Matcher<? super T>
Errors: 1. Warnings: 0.
```

The full Travis log is here: https://travis-ci.org/facebook/buck/builds/189437539 .

**Comment by cushon** [comment]
That code doesn't compile with a stock JDK 9, this isn't specific to Error Prone. I think it's related to [JDK-8039214](https://bugs.openjdk.java.net/browse/JDK-8039214).

If you want a work-around, `return hasConsecutiveItems(Arrays.<Matcher<? super T>>asList(matchers));` should do it.

Repro:

```java
import java.util.*;

class Test {
  interface Matcher<T> {}

  public static <T> Matcher<Collection<? extends T>> hasConsecutiveItems(
      Matcher<? super T>... matchers) {
    return hasConsecutiveItems(Arrays.asList(matchers));
  }

  public static <T> Matcher<Collection<? extends T>> hasConsecutiveItems(T... elements) {
    return hasConsecutiveItems(Arrays.asList(elements));
  }

  public static <T> Matcher<Collection<? extends T>> hasConsecutiveItems(
      List<Matcher<? super T>> matchers) {
    throw new AssertionError();
  }

  public static <T> Matcher<Collection<? extends T>> hasConsecutiveItems(
      Iterable<? extends T> items) {
    throw new AssertionError();
  }
}
```

```
$ javac -fullversion Test.java
javac full version "1.8.0_122-ea-b04"
... OK
$ javac -fullversion Test.java
javac full version "9-ea+151"
Test.java:8: error: incompatible types: inference variable T#1 has incompatible bounds
    return hasConsecutiveItems(Arrays.asList(matchers));
                              ^
    equality constraints: T#2
    lower bounds: T#3,Matcher<? super T#2>
  where T#1,T#2,T#3 are type-variables:
    T#1 extends Object declared in method <T#1>hasConsecutiveItems(Iterable<? extends T#1>)
    T#2 extends Object declared in method <T#2>hasConsecutiveItems(Matcher<? super T#2>...)
    T#3 extends Object declared in method <T#3>asList(T#3...)
```

**Comment by davido** [comment]
Thanks for clarification and for suggested workaround. It works.
